### PR TITLE
use is_serialized() instead of is_string()

### LINKED ISSUE #122: Warning: unserialize() expects parameter 1 to be string, array given
## Bug Report

- [ ] Yes, I reviewed the [contribution guidelines](https://make.wordpress.org/cli/handbook/contributing/).
- [ ] Yes, more specifically, I reviewed the guidelines on [how to write clear bug reports](https://make.wordpress.org/cli/handbook/bug-reports/).

**Describe the current, buggy behavior**

```
[Suppressed] Warning: unserialize() expects parameter 1 to be string, array given in wp-cli/search-replace-command/src/WP_CLI/SearchReplacer.php on line 85
```

Additionally, this results in warnings like this:

```
[Suppressed] Notice: unserialize(): Error at offset 0 of 24 bytes in wp-cli/search-replace-command/src/WP_CLI/SearchReplacer.php on line 85
```

**Describe how other contributors can replicate this bug**

I'm not 100% sure the conditions that trigger this, we've just noticed it while running search/replaces like this:

```
wp --path=/var/www/ search-replace http://foo.com https://foo.bar --all-tables --dry-run
```

**Describe what you would expect as the correct outcome**

No suppressed warnings because only strings get passed into `unserialize()`.

**Let us know what environment you are running this on**

```
wp cli info
OS:    Linux 4.9.0-9-amd64 #1 SMP Debian 4.9.168-1 (2019-04-12) x86_64
Shell:    /bin/bash
PHP binary:    /<redacted>/php
PHP version:    7.3.5
php.ini used:    /<redacted>/php.ini
WP-CLI root dir:    phar://wp-cli.phar/vendor/wp-cli/wp-cli
WP-CLI vendor dir:    phar://wp-cli.phar/vendor
WP_CLI phar path:    /home/<redacted>
WP-CLI packages dir:    
WP-CLI global config:    
WP-CLI project config:    
WP-CLI version:    2.1.0
```

**Provide a possible solution**

This bug was introduced here:

https://github.com/wp-cli/search-replace-command/commit/babbfbfcc1d36aff2f8b54f7b491bded98bde5d3

It would seem that while trying to address PHPCS styles, the logic was changed.

Was:

```
if ( is_string( $data ) && false !== ( $unserialized = @unserialize( $data ) ) ) {
```

Is now:

```
$unserialized = @unserialize( $data );
if ( is_string( $data ) && false !== $unserialized ) {
```

The `is_string()` check needs to happen _before_ `unserialize()` is called, not after.

### PR DESCRIPTION
**Pull request by ideag**
Fixes #122 - serialization warnings in PHP 8.*

### CODE REVIEW

**Comment by rwagner00** [comment]
Will confirm that it resolved https://github.com/wp-cli/search-replace-command/issues/122 for me and works as intended.

File: src/WP_CLI/SearchReplacer.php [hunk 1]
```diff
@@ -83,7 +83,7 @@ private function run_recursively( $data, $serialised, $recursion_level = 0, $vis
 				}
 			}
 
-			$unserialized = is_string( $data ) ? @unserialize( $data ) : false;
+			$unserialized = is_serialized( $data ) ? @unserialize( $data ) : false;
```
Comment (danielbachhuber) [inline src/WP_CLI/SearchReplacer.php:86]: This generally seems fine. Could we get a test case for it?

Also, let's internalize `is_serialized()` to a private method on this class. As far as I can tell, we don't have any other WordPress-specific dependencies in this class, so better to keep things that way for now.

**Review by danielbachhuber — CHANGES_REQUESTED** [review]

File: src/WP_CLI/SearchReplacer.php (hunk 1 above)
Comment (schlessera) [reply src/WP_CLI/SearchReplacer.php:86]: Yes, during a `search-replace`, WP might potentially be broken, so it's best to copy the function over.

**Comment by danielbachhuber** [comment]
Proceeding with https://github.com/wp-cli/wp-cli/issues/5594 for this repository. I've captured this PR to https://gist.github.com/danielbachhuber/6c429bfdd43cae3ddb72559ad1f54446 in case this PR is auto-closed or broken in some way.

**Pull request closed without merging**